geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/ReplayKafkaConsumerFeatureSource.scala

```diff
@@ -0,0 +1,419 @@
+    ReplayConfig(new Instant(start), new Instant(end), Duration.millis(readBehind))
+
+  def encode(conf: ReplayConfig): String =
+    s"${conf.start.getMillis.toHexString}-${conf.end.getMillis.toHexString}-${conf.readBehind.getMillis.toHexString}"
+
+  def decode(rcString: String): Option[ReplayConfig] = {
+
+    try {
+      val values = rcString.split('-').map(java.lang.Long.valueOf(_, 16))
+
+      if (values.length != 3) {
+        logger.error("Unable to decode ReplayConfig. Wrong number of tokens splitting " + rcString)
+        None
+      } else {
+
+        val start = new Instant(values(0))
+        val end = new Instant(values(1))
+        val duration = Duration.millis(values(2))
+        Some(ReplayConfig(start, end, duration))
+      }
+    } catch {
+      case e: IllegalArgumentException  =>
+        logger.error("Exception thrown decoding ReplayConfig.", e)
+        None
+    }
+  }
+}
+
+/** Splits a [[Filter]] into the requested Kafka Message Timestamp and the remaining filters
+  */
+case class TimestampFilterSplit(ts: Option[Long], filter: Option[Filter])
+
+object TimestampFilterSplit {
+
+  import ReplayTimeHelper.ff
+
+  /** Look for a Kafka message timestamp filter in ``filter`` and if found, extract the requested timestamp
+    * and return that timestamp and the remaining filters.
+    *
+    * Any operand (or none) of an 'and' may specify a timestamp.  If multiple operands of the 'and'
+    * specify a timestamp then all timestamps must be the same.
+    *
+    * For an 'or' the requirement is that either all operands specify the same timestamp or none specify a
+    * timestamp.
+    *
+    * A timestamp may not be specified within a 'not'.
+    */
+  def split(filter: Filter): Option[TimestampFilterSplit] = filter match {
+
+    case eq: PropertyIsEqualTo =>
+      val ts = ReplayTimeHelper.fromFilter(eq)
+      val f = ts.map(_ => None).getOrElse(Some(filter))
+      Some(TimestampFilterSplit(ts, f))
+
+    case a: And =>
+      // either no child specifies a timestamp, one child specifies a timestamp or multiple children specify
+      // the same timestamp
+      split(a, buildAnd)
+
+    case o: Or =>
+      // either all children specify the same timestamp or none specify a timestamp
+      split(o, buildOr)
+
+    case n: Not =>
+      // the filter being inverted may not contain a timestamp
+      val s = split(n.getFilter)
+      s.flatMap(split => split.ts.map(_ => None)
+        .getOrElse(Some(TimestampFilterSplit(None, split.filter.map(ff.not)))))
+
+    case _ => Some(TimestampFilterSplit(None, Some(filter)))
+  }
+
+  type SplitCombiner = Seq[TimestampFilterSplit] => Option[TimestampFilterSplit]
+
+  def split(op: BinaryLogicOperator, combiner: SplitCombiner): Option[TimestampFilterSplit] = {
+    val children = op.getChildren.asScala
+    val childSplits = children.flatMap(c => split(c))
+
+    if (childSplits.size != children.size) {
+      // one or more children are invalid
```

Comment:
I could handle each side of the OR and then union the results.  But, you might end up with multiple different versions of the same simple feature at different points in time.  There might be a use case for this but there is also an issues with the different versions having the same ID but not being equal.  Currently we don't need to support this, as far as I know.